File: app/src/main/java/com/damianmichalak/shopping_list/model/CurrentListDao.java
package com.damianmichalak.shopping_list.model;


import com.damianmichalak.shopping_list.helper.References;
import com.damianmichalak.shopping_list.helper.EventsWrapper;
import com.damianmichalak.shopping_list.helper.RxUtils;
import com.damianmichalak.shopping_list.model.api_models.ShoppingList;

import javax.annotation.Nonnull;
import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.Observer;
import rx.observers.Observers;
import rx.subjects.PublishSubject;

@Singleton
public class CurrentListDao {

    @Nonnull
    private final Observable<ShoppingList> currentListObservable;
    @Nonnull
    private final Observable<String> currentListKeyObservable;
    @Nonnull
    private final UserPreferences userPreferences;
    @Nonnull
    private final References References;
    @Nonnull
    private final PublishSubject<Object> currentListRefreshSubject = PublishSubject.create();

    @Inject
    public CurrentListDao(@Nonnull final UserPreferences userPreferences,
                          @Nonnull final References References,
                          @Nonnull final EventsWrapper wrapper) {
        this.userPreferences = userPreferences;
        this.References = References;

        currentListObservable = currentListRefreshSubject.startWith(((Object) null))
                .flatMap(o -> Observable.fromCallable(userPreferences::getCurrentList))
                .filter(uid -> uid != null)
                .flatMap(uid -> RxUtils.createObservableForReference(References.singleListReference(uid), wrapper, ShoppingList.class))
                .replay(1)
                .refCount();

        currentListKeyObservable = currentListRefreshSubject.startWith(((Object) null))
                .flatMap(o -> Observable.fromCallable(userPreferences::getCurrentList))
                .filter(uid -> uid != null);

    }

    @Nonnull
    public Observable<String> getCurrentListKeyObservable() {
        return currentListKeyObservable;
    }

    @Nonnull
    public Observable<ShoppingList> getCurrentListObservable() {
        return currentListObservable;
    }

    @Nonnull
    public Observer<String> saveCurrentListIdObserver() {
        return Observers.create(currentListID1 -> {
            userPreferences.setCurrentList(currentListID1);
            currentListRefreshSubject.onNext(null);
        });
    }
}

File: app/src/main/java/com/damianmichalak/shopping_list/model/CurrentListDao.kt
package com.damianmichalak.shopping_list.model


import com.damianmichalak.shopping_list.helper.EventsWrapper
import com.damianmichalak.shopping_list.helper.References
import com.damianmichalak.shopping_list.helper.RxUtils
import com.damianmichalak.shopping_list.model.apiModels.ShoppingList
import rx.Observable
import rx.Observer
import rx.observers.Observers
import rx.subjects.PublishSubject
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CurrentListDao @Inject constructor(private val userPreferences: UserPreferences,
                                         references: References,
                                         wrapper: EventsWrapper) {

    val currentListObservable: Observable<ShoppingList>
    val currentListKeyObservable: Observable<String>
    private val currentListRefreshSubject = PublishSubject.create<Any>()

    init {

        currentListObservable = currentListRefreshSubject.startWith(null as Any?)
                .flatMap { o -> Observable.fromCallable { userPreferences.currentList } }
                .filter { uid -> uid != null }
                .flatMap { uid -> RxUtils.createObservableForReference(references.singleListReference(uid), wrapper, ShoppingList::class.java) }
                .replay(1)
                .refCount()

        currentListKeyObservable = currentListRefreshSubject.startWith(null as Any?)
                .flatMap { o -> Observable.fromCallable { userPreferences.currentList } }
                .filter { uid -> uid != null }
                .map { it }

    }

    fun saveCurrentListIdObserver(): Observer<String> {
        return Observers.create { currentListID1 ->
            userPreferences.currentList = currentListID1
            currentListRefreshSubject.onNext(null)
        }
    }
}
